VlsirTools/vlsirtools/spice/ngspice.py

```diff
@@ -0,0 +1,452 @@
+        )
+
+    def parse_op(self, an: vsp.OpInput, nutbin: "NutBinAnalysis") -> OpResult:
+        return OpResult(
+            analysis_name=an.analysis_name,
+            data={k: v[0] for k, v in nutbin.data.items()},
+        )
+
+    def parse_tran(self, an: vsp.TranInput, nutbin: "NutBinAnalysis") -> TranResult:
+        """ Extract the results for Analysis `an` from `data`. """
+        measurements = self.get_measurements("*.mt*")
+        return TranResult(
+            analysis_name=an.analysis_name, data=nutbin.data, measurements=measurements
+        )
+
+    def get_measurements(self, filepat: str) -> Dict[str, float]:
+        """ Get the measurements at files matching (glob) `filepat`. 
+        Returns only a single files-worth of measurements, and issues a warning if more than one such file exists. 
+        Returns an empty dictionary if no matching files are found. """
+        meas_files = list(self.glob(filepat))
+        if not meas_files:
+            return dict()
+        if len(meas_files) > 1:
+            msg = f"Unsupported: more than one measurement-file generated. Only the first will be read"
+            warn(msg)
+        return parse_mt0(self.open(meas_files[0], "r"))
+
+    def run_spectre_process(self) -> Awaitable[None]:
+        """ Run a NGSpice sub-process, executing the simulation """
+        # Note the `nutbin` output format is dictated here
+        cmd = f"{NGSPICE_EXECUTABLE} -b netlist.sp -r netlist.raw"
+        return self.run_subprocess(cmd)
+
+
+class FromStr:
+    """ Mix-in for creating `Enum`s from string values. """
+
+    @classmethod 
+    def from_str(cls, s: str) -> "FromStr":
+        reversed = {v.value: v for v in cls}
+        if s not in reversed:
+            msg = f"Invalid `{cls.__name__}`: `{s}`"
+            raise ValueError(msg)
+        return reversed[s]
+
+
+class NumType(FromStr, Enum):
+    """ # Numeric Datatypes 
+    Values equal the strings used in NutBin data files. """
+
+    REAL = "real"
+    COMPLEX = "complex"
+
+
+class Units(FromStr, Enum):
+    """ # Unit Types
+    Values equal the strings used in NutBin data files. """
+
+    DUMMY = "dummy"
+    SWEEP = "sweep"
+    SECONDS = "seconds"
+    VOLTS = "voltage"
+    AMPS = "current"
+    HERTZ = "frequency"
+    CELSIUS = "C"
+
+    @classmethod
+    def from_str(cls, s: str) -> "FromStr":
+        reversed_l = {v.value: v for v in cls}
+        if s not in reversed_l:
+            return cls.DUMMY
+        return reversed_l[s]
+
+
+@dataclass
+class VarSpec:
+    """ Variable Spec """
+    
+    name: str 
+    units: Units
+
+
+@dataclass 
+class NutBinAnalysis:
+    """ Analysis result from a NutBin file. """
+
+    analysis_name: str  # Analysis name
+    numtype: NumType  # Numeric type in `data` field
+    data: Mapping[str, np.ndarray]  # Signal name => data
+    units: Mapping[str, Units]  # Signal name => units
+
+
+def parse_nutbin(f: IO) -> Mapping[str, NutBinAnalysis]:
```

Comment:
Is all of this `nutbin` stuff identical to the version copied over from `xyce.py`? 
Or did we have to customize it at all?

Reply:
I built this from spectre.py, and it did require some changes. The major one was with how spectre adds an extra new line character in some places while ngspice does not. It could be that this lines up with xyce, I did not check